// @researchable/fitbit-web-api/src/api/ActivityApi.js
/**
 * Fitbit Web API
 * Fitbit provides a Web API for accessing data from Fitbit activity trackers, Aria scale, and manually entered logs. Anyone can develop an application to access and modify a Fitbit user's data on their behalf, so long as it complies with Fitbit Platform Terms of Service. These Swagger UI docs do not currently support making Fitbit API requests directly. In order to make a request, construct a request for the appropriate endpoint using this documentation, and then add an Authorization header to each request with an access token obtained using the steps outlined here: https://dev.fitbit.com/build/reference/web-api/oauth2/#obtaining-consent.
 *
 * OpenAPI spec version: 1
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.4.7
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.FitbitWebApi) {
      root.FitbitWebApi = {};
    }
    root.FitbitWebApi.ActivityApi = factory(root.FitbitWebApi.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';

  /**
   * Activity service.
   * @module api/ActivityApi
   * @version 1
   */

  /**
   * Constructs a new ActivityApi. 
   * @alias module:api/ActivityApi
   * @class
   * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
   * default to {@link module:ApiClient#instance} if unspecified.
   */
  var exports = function(apiClient) {
    this.apiClient = apiClient || ApiClient.instance;


    /**
     * Callback function to receive the result of the addActivitiesLog operation.
     * @callback module:api/ActivityApi~addActivitiesLogCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Log Activity
     * The Log Activity endpoint creates log entry for an activity or user&#39;s private custom activity using units in the unit system which corresponds to the Accept-Language header provided (or using optional custom distanceUnit) and get a response in the format requested.
     * @param {Number} activityId The ID of the activity, directory activity or intensity level activity.
     * @param {Number} manualCalories Calories burned that are manaully specified. Required with activityName must be provided.
     * @param {String} startTime Activity start time. Hours and minutes in the format HH:mm:ss.
     * @param {Number} durationMillis Duration in milliseconds.
     * @param {Date} _date Log entry date in the format yyyy-MM-dd.
     * @param {Number} distance Distance is required for logging directory activity in the format X.XX and in the selected distanceUnit.
     * @param {Object} opts Optional parameters
     * @param {String} opts.activityName Custom activity name. Either activityId or activityName must be provided.
     * @param {Number} opts.distanceUnit Distance measurement unit. Steps units are available only for Walking (activityId&#x3D;90013) and Running (activityId&#x3D;90009) directory activities and their intensity levels.
     * @param {module:api/ActivityApi~addActivitiesLogCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.addActivitiesLog = function(activityId, manualCalories, startTime, durationMillis, _date, distance, opts, callback) {
      opts = opts || {};
      var postBody = null;

      // verify the required parameter 'activityId' is set
      if (activityId === undefined || activityId === null) {
        throw new Error("Missing the required parameter 'activityId' when calling addActivitiesLog");
      }

      // verify the required parameter 'manualCalories' is set
      if (manualCalories === undefined || manualCalories === null) {
        throw new Error("Missing the required parameter 'manualCalories' when calling addActivitiesLog");
      }

      // verify the required parameter 'startTime' is set
      if (startTime === undefined || startTime === null) {
        throw new Error("Missing the required parameter 'startTime' when calling addActivitiesLog");
      }

      // verify the required parameter 'durationMillis' is set
      if (durationMillis === undefined || durationMillis === null) {
        throw new Error("Missing the required parameter 'durationMillis' when calling addActivitiesLog");
      }

      // verify the required parameter '_date' is set
      if (_date === undefined || _date === null) {
        throw new Error("Missing the required parameter '_date' when calling addActivitiesLog");
      }

      // verify the required parameter 'distance' is set
      if (distance === undefined || distance === null) {
        throw new Error("Missing the required parameter 'distance' when calling addActivitiesLog");
      }


      var pathParams = {
      };
      var queryParams = {
        'activityId': activityId,
        'activityName': opts['activityName'],
        'manualCalories': manualCalories,
        'startTime': startTime,
        'durationMillis': durationMillis,
        'date': _date,
        'distance': distance,
        'distanceUnit': opts['distanceUnit'],
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities.json', 'POST',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the addFavoriteActivities operation.
     * @callback module:api/ActivityApi~addFavoriteActivitiesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Add Favorite Activity
     * Adds the activity with the given ID to user&#39;s list of favorite activities.
     * @param {String} activityId The encoded ID of the activity.
     * @param {module:api/ActivityApi~addFavoriteActivitiesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.addFavoriteActivities = function(activityId, callback) {
      var postBody = null;

      // verify the required parameter 'activityId' is set
      if (activityId === undefined || activityId === null) {
        throw new Error("Missing the required parameter 'activityId' when calling addFavoriteActivities");
      }


      var pathParams = {
        'activity-id': activityId
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/favorite/{activity-id}.json', 'POST',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the addUpdateActivitiesGoals operation.
     * @callback module:api/ActivityApi~addUpdateActivitiesGoalsCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Update Activity Goals
     * Updates a user&#39;s daily or weekly activity goals and returns a response using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {String} period daily or weekly.
     * @param {String} type goal type
     * @param {String} value goal value
     * @param {module:api/ActivityApi~addUpdateActivitiesGoalsCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.addUpdateActivitiesGoals = function(period, type, value, callback) {
      var postBody = null;

      // verify the required parameter 'period' is set
      if (period === undefined || period === null) {
        throw new Error("Missing the required parameter 'period' when calling addUpdateActivitiesGoals");
      }

      // verify the required parameter 'type' is set
      if (type === undefined || type === null) {
        throw new Error("Missing the required parameter 'type' when calling addUpdateActivitiesGoals");
      }

      // verify the required parameter 'value' is set
      if (value === undefined || value === null) {
        throw new Error("Missing the required parameter 'value' when calling addUpdateActivitiesGoals");
      }


      var pathParams = {
        'period': period
      };
      var queryParams = {
        'type': type,
        'value': value,
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/goals/{period}.json', 'POST',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the deleteActivitiesLog operation.
     * @callback module:api/ActivityApi~deleteActivitiesLogCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Delete Activity Log
     * Deletes a user&#39;s activity log entry with the given ID.
     * @param {Number} activityLogId The id of the activity log entry.
     * @param {module:api/ActivityApi~deleteActivitiesLogCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.deleteActivitiesLog = function(activityLogId, callback) {
      var postBody = null;

      // verify the required parameter 'activityLogId' is set
      if (activityLogId === undefined || activityLogId === null) {
        throw new Error("Missing the required parameter 'activityLogId' when calling deleteActivitiesLog");
      }


      var pathParams = {
        'activity-log-id': activityLogId
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/{activity-log-id}.json', 'DELETE',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the deleteFavoriteActivities operation.
     * @callback module:api/ActivityApi~deleteFavoriteActivitiesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Delete Favorite Activity
     * Removes the activity with the given ID from a user&#39;s list of favorite activities.
     * @param {String} activityId The ID of the activity to be removed.
     * @param {module:api/ActivityApi~deleteFavoriteActivitiesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.deleteFavoriteActivities = function(activityId, callback) {
      var postBody = null;

      // verify the required parameter 'activityId' is set
      if (activityId === undefined || activityId === null) {
        throw new Error("Missing the required parameter 'activityId' when calling deleteFavoriteActivities");
      }


      var pathParams = {
        'activity-id': activityId
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/favorite/{activity-id}.json', 'DELETE',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesByDate operation.
     * @callback module:api/ActivityApi~getActivitiesByDateCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Activity Summary by Date 
     * Retrieves a summary and list of a user&#39;s activities and activity log entries for a given day.
     * @param {Date} _date The date in the format yyyy-MM-dd
     * @param {module:api/ActivityApi~getActivitiesByDateCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesByDate = function(_date, callback) {
      var postBody = null;

      // verify the required parameter '_date' is set
      if (_date === undefined || _date === null) {
        throw new Error("Missing the required parameter '_date' when calling getActivitiesByDate");
      }


      var pathParams = {
        'date': _date
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/date/{date}.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesGoals operation.
     * @callback module:api/ActivityApi~getActivitiesGoalsCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Activity Goals
     * Retreives a user&#39;s current daily or weekly activity goals using measurement units as defined in the unit system, which corresponds to the Accept-Language header provided.
     * @param {String} period daily or weekly.
     * @param {module:api/ActivityApi~getActivitiesGoalsCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesGoals = function(period, callback) {
      var postBody = null;

      // verify the required parameter 'period' is set
      if (period === undefined || period === null) {
        throw new Error("Missing the required parameter 'period' when calling getActivitiesGoals");
      }


      var pathParams = {
        'period': period
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/goals/{period}.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesLog operation.
     * @callback module:api/ActivityApi~getActivitiesLogCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Lifetime Stats
     * Updates a user&#39;s daily activity goals and returns a response using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {module:api/ActivityApi~getActivitiesLogCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesLog = function(callback) {
      var postBody = null;


      var pathParams = {
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesLogList operation.
     * @callback module:api/ActivityApi~getActivitiesLogListCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Activity Log List
     * Retreives a list of user&#39;s activity log entries before or after a given day with offset and limit using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {String} sort The sort order of entries by date asc (ascending) or desc (descending).
     * @param {Number} offset The offset number of entries.
     * @param {Number} limit The maximum number of entries returned (maximum;100).
     * @param {Object} opts Optional parameters
     * @param {Date} opts.beforeDate The date in the format yyyy-MM-ddTHH:mm:ss. Only yyyy-MM-dd is required. Either beforeDate or afterDate should be specified.
     * @param {Date} opts.afterDate The date in the format yyyy-MM-ddTHH:mm:ss.
     * @param {module:api/ActivityApi~getActivitiesLogListCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesLogList = function(sort, offset, limit, opts, callback) {
      opts = opts || {};
      var postBody = null;

      // verify the required parameter 'sort' is set
      if (sort === undefined || sort === null) {
        throw new Error("Missing the required parameter 'sort' when calling getActivitiesLogList");
      }

      // verify the required parameter 'offset' is set
      if (offset === undefined || offset === null) {
        throw new Error("Missing the required parameter 'offset' when calling getActivitiesLogList");
      }

      // verify the required parameter 'limit' is set
      if (limit === undefined || limit === null) {
        throw new Error("Missing the required parameter 'limit' when calling getActivitiesLogList");
      }


      var pathParams = {
      };
      var queryParams = {
        'beforeDate': opts['beforeDate'],
        'afterDate': opts['afterDate'],
        'sort': sort,
        'offset': offset,
        'limit': limit,
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/list.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesTCX operation.
     * @callback module:api/ActivityApi~getActivitiesTCXCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Activity TCX
     * Retreives the details of a user&#39;s location and heart rate data during a logged exercise activity.
     * @param {String} logId The activity&#39;s log ID.
     * @param {Object} opts Optional parameters
     * @param {Boolean} opts.includePartialTCX Include TCX points regardless of GPS data being present
     * @param {module:api/ActivityApi~getActivitiesTCXCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesTCX = function(logId, opts, callback) {
      opts = opts || {};
      var postBody = null;

      // verify the required parameter 'logId' is set
      if (logId === undefined || logId === null) {
        throw new Error("Missing the required parameter 'logId' when calling getActivitiesTCX");
      }


      var pathParams = {
        'log-id': logId
      };
      var queryParams = {
        'includePartialTCX': opts['includePartialTCX'],
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/{log-id}.tcx', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesTypeDetail operation.
     * @callback module:api/ActivityApi~getActivitiesTypeDetailCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Activity Type
     * Returns the detail of a specific activity in the Fitbit activities database in the format requested. If activity has levels, it also returns a list of activity level details.
     * @param {String} activityId The activity ID.
     * @param {module:api/ActivityApi~getActivitiesTypeDetailCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesTypeDetail = function(activityId, callback) {
      var postBody = null;

      // verify the required parameter 'activityId' is set
      if (activityId === undefined || activityId === null) {
        throw new Error("Missing the required parameter 'activityId' when calling getActivitiesTypeDetail");
      }


      var pathParams = {
        'activity-id': activityId
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/activities/{activity-id}.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getActivitiesTypes operation.
     * @callback module:api/ActivityApi~getActivitiesTypesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Browse Activity Types
     * Retreives a tree of all valid Fitbit public activities from the activities catelog as well as private custom activities the user created in the format requested.
     * @param {module:api/ActivityApi~getActivitiesTypesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getActivitiesTypes = function(callback) {
      var postBody = null;


      var pathParams = {
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/activities.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getFavoriteActivities operation.
     * @callback module:api/ActivityApi~getFavoriteActivitiesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Favorite Activities
     * Returns a list of a user&#39;s favorite activities.
     * @param {module:api/ActivityApi~getFavoriteActivitiesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getFavoriteActivities = function(callback) {
      var postBody = null;


      var pathParams = {
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/favorite.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getFrequentActivities operation.
     * @callback module:api/ActivityApi~getFrequentActivitiesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Frequent Activities
     * Retreives a list of a user&#39;s frequent activities in the format requested using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {module:api/ActivityApi~getFrequentActivitiesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getFrequentActivities = function(callback) {
      var postBody = null;


      var pathParams = {
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/frequent.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getRecentActivities operation.
     * @callback module:api/ActivityApi~getRecentActivitiesCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Recent Activity Types
     * Retreives a list of a user&#39;s recent activities types logged with some details of the last activity log of that type using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {module:api/ActivityApi~getRecentActivitiesCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getRecentActivities = function(callback) {
      var postBody = null;


      var pathParams = {
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/x-www-form-urlencoded'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/activities/recent.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }
  };

  return exports;
}));
