// @researchable/fitbit-web-api/src/api/FoodAndWaterTimeSeriesApi.js
/**
 * Fitbit Web API
 * Fitbit provides a Web API for accessing data from Fitbit activity trackers, Aria scale, and manually entered logs. Anyone can develop an application to access and modify a Fitbit user's data on their behalf, so long as it complies with Fitbit Platform Terms of Service. These Swagger UI docs do not currently support making Fitbit API requests directly. In order to make a request, construct a request for the appropriate endpoint using this documentation, and then add an Authorization header to each request with an access token obtained using the steps outlined here: https://dev.fitbit.com/build/reference/web-api/oauth2/#obtaining-consent.
 *
 * OpenAPI spec version: 1
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.4.7
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.FitbitWebApi) {
      root.FitbitWebApi = {};
    }
    root.FitbitWebApi.FoodAndWaterTimeSeriesApi = factory(root.FitbitWebApi.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';

  /**
   * FoodAndWaterTimeSeries service.
   * @module api/FoodAndWaterTimeSeriesApi
   * @version 1
   */

  /**
   * Constructs a new FoodAndWaterTimeSeriesApi. 
   * @alias module:api/FoodAndWaterTimeSeriesApi
   * @class
   * @param {module:ApiClient} [apiClient] Optional API client implementation to use,
   * default to {@link module:ApiClient#instance} if unspecified.
   */
  var exports = function(apiClient) {
    this.apiClient = apiClient || ApiClient.instance;


    /**
     * Callback function to receive the result of the getFoodsByDateRange operation.
     * @callback module:api/FoodAndWaterTimeSeriesApi~getFoodsByDateRangeCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Food or Water Time Series
     * Updates a user&#39;s daily activity goals and returns a response using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {module:model/String} resourcePath The resouce path. See options in the Resouce Path Options section in the full documentation.
     * @param {Date} baseDate The range start date in the format yyyy-MM-dd or today.
     * @param {Date} endDate The end date of the range.
     * @param {module:api/FoodAndWaterTimeSeriesApi~getFoodsByDateRangeCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getFoodsByDateRange = function(resourcePath, baseDate, endDate, callback) {
      var postBody = null;

      // verify the required parameter 'resourcePath' is set
      if (resourcePath === undefined || resourcePath === null) {
        throw new Error("Missing the required parameter 'resourcePath' when calling getFoodsByDateRange");
      }

      // verify the required parameter 'baseDate' is set
      if (baseDate === undefined || baseDate === null) {
        throw new Error("Missing the required parameter 'baseDate' when calling getFoodsByDateRange");
      }

      // verify the required parameter 'endDate' is set
      if (endDate === undefined || endDate === null) {
        throw new Error("Missing the required parameter 'endDate' when calling getFoodsByDateRange");
      }


      var pathParams = {
        'resource-path': resourcePath,
        'base-date': baseDate,
        'end-date': endDate
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/json'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/foods/log/{resource-path}/date/{base-date}/{end-date}.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }

    /**
     * Callback function to receive the result of the getFoodsResourceByDatePeriod operation.
     * @callback module:api/FoodAndWaterTimeSeriesApi~getFoodsResourceByDatePeriodCallback
     * @param {String} error Error message, if any.
     * @param data This operation does not return a value.
     * @param {String} response The complete HTTP response.
     */

    /**
     * Get Food or Water Time Series
     * Updates a user&#39;s daily activity goals and returns a response using units in the unit system which corresponds to the Accept-Language header provided.
     * @param {module:model/String} resourcePath The resouce path. See options in the Resouce Path Options section in the full documentation.
     * @param {Date} _date The end date of the period specified in the format yyyy-MM-dd or today.
     * @param {String} period The range for which data will be returned. Options are 1d, 7d, 30d, 1w, 3m, 6m, 1y, or max.
     * @param {module:api/FoodAndWaterTimeSeriesApi~getFoodsResourceByDatePeriodCallback} callback The callback function, accepting three arguments: error, data, response
     */
    this.getFoodsResourceByDatePeriod = function(resourcePath, _date, period, callback) {
      var postBody = null;

      // verify the required parameter 'resourcePath' is set
      if (resourcePath === undefined || resourcePath === null) {
        throw new Error("Missing the required parameter 'resourcePath' when calling getFoodsResourceByDatePeriod");
      }

      // verify the required parameter '_date' is set
      if (_date === undefined || _date === null) {
        throw new Error("Missing the required parameter '_date' when calling getFoodsResourceByDatePeriod");
      }

      // verify the required parameter 'period' is set
      if (period === undefined || period === null) {
        throw new Error("Missing the required parameter 'period' when calling getFoodsResourceByDatePeriod");
      }


      var pathParams = {
        'resource-path': resourcePath,
        'date': _date,
        'period': period
      };
      var queryParams = {
      };
      var collectionQueryParams = {
      };
      var headerParams = {
      };
      var formParams = {
      };

      var authNames = ['oauth2'];
      var contentTypes = [];
      var accepts = ['application/json'];
      var returnType = null;

      return this.apiClient.callApi(
        '/1/user/-/foods/log/{resource-path}/date/{date}/{period}.json', 'GET',
        pathParams, queryParams, collectionQueryParams, headerParams, formParams, postBody,
        authNames, contentTypes, accepts, returnType, callback
      );
    }
  };

  return exports;
}));
