// @researchable/fitbit-web-api/src/index.js
/**
 * Fitbit Web API
 * Fitbit provides a Web API for accessing data from Fitbit activity trackers, Aria scale, and manually entered logs. Anyone can develop an application to access and modify a Fitbit user's data on their behalf, so long as it complies with Fitbit Platform Terms of Service. These Swagger UI docs do not currently support making Fitbit API requests directly. In order to make a request, construct a request for the appropriate endpoint using this documentation, and then add an Authorization header to each request with an access token obtained using the steps outlined here: https://dev.fitbit.com/build/reference/web-api/oauth2/#obtaining-consent.
 *
 * OpenAPI spec version: 1
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.4.7
 *
 * Do not edit the class manually.
 *
 */

(function(factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient', 'api/ActivityApi', 'api/ActivityIntradayTimeSeriesApi', 'api/ActivityTimeSeriesApi', 'api/AuthApi', 'api/BodyAndWeightApi', 'api/BodyAndWeightTimeSeriesApi', 'api/DevicesApi', 'api/FoodAndWaterApi', 'api/FoodAndWaterTimeSeriesApi', 'api/FriendsApi', 'api/HeartRateIntradayTimeSeriesApi', 'api/HeartRateTimeSeriesApi', 'api/SleepApi', 'api/SubscriptionsApi', 'api/UserApi'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('./ApiClient'), require('./api/ActivityApi'), require('./api/ActivityIntradayTimeSeriesApi'), require('./api/ActivityTimeSeriesApi'), require('./api/AuthApi'), require('./api/BodyAndWeightApi'), require('./api/BodyAndWeightTimeSeriesApi'), require('./api/DevicesApi'), require('./api/FoodAndWaterApi'), require('./api/FoodAndWaterTimeSeriesApi'), require('./api/FriendsApi'), require('./api/HeartRateIntradayTimeSeriesApi'), require('./api/HeartRateTimeSeriesApi'), require('./api/SleepApi'), require('./api/SubscriptionsApi'), require('./api/UserApi'));
  }
}(function(ApiClient, ActivityApi, ActivityIntradayTimeSeriesApi, ActivityTimeSeriesApi, AuthApi, BodyAndWeightApi, BodyAndWeightTimeSeriesApi, DevicesApi, FoodAndWaterApi, FoodAndWaterTimeSeriesApi, FriendsApi, HeartRateIntradayTimeSeriesApi, HeartRateTimeSeriesApi, SleepApi, SubscriptionsApi, UserApi) {
  'use strict';

  /**
   * Fitbit_provides_a_Web_API_for_accessing_data_from_Fitbit_activity_trackers_Aria_scale_and_manually_entered_logs__Anyone_can_develop_an_application_to_access_and_modify_a_Fitbit_users_data_on_their_behalf_so_long_as_it_complies_with_Fitbit_Platform_Terms_of_Service__These_Swagger_UI_docs_do_not_currently_support_making_Fitbit_API_requests_directly__In_order_to_make_a_request_construct_a_request_for_the_appropriate_endpoint_using_this_documentation_and_then_add_an_Authorization_header_to_each_request_with_an_access_token_obtained_using_the_steps_outlined_here_httpsdev_fitbit_combuildreferenceweb_apioauth2obtaining_consent_.<br>
   * The <code>index</code> module provides access to constructors for all the classes which comprise the public API.
   * <p>
   * An AMD (recommended!) or CommonJS application will generally do something equivalent to the following:
   * <pre>
   * var FitbitWebApi = require('index'); // See note below*.
   * var xxxSvc = new FitbitWebApi.XxxApi(); // Allocate the API class we're going to use.
   * var yyyModel = new FitbitWebApi.Yyy(); // Construct a model instance.
   * yyyModel.someProperty = 'someValue';
   * ...
   * var zzz = xxxSvc.doSomething(yyyModel); // Invoke the service.
   * ...
   * </pre>
   * <em>*NOTE: For a top-level AMD script, use require(['index'], function(){...})
   * and put the application logic within the callback function.</em>
   * </p>
   * <p>
   * A non-AMD browser application (discouraged) might do something like this:
   * <pre>
   * var xxxSvc = new FitbitWebApi.XxxApi(); // Allocate the API class we're going to use.
   * var yyy = new FitbitWebApi.Yyy(); // Construct a model instance.
   * yyyModel.someProperty = 'someValue';
   * ...
   * var zzz = xxxSvc.doSomething(yyyModel); // Invoke the service.
   * ...
   * </pre>
   * </p>
   * @module index
   * @version 1
   */
  var exports = {
    /**
     * The ApiClient constructor.
     * @property {module:ApiClient}
     */
    ApiClient: ApiClient,
    /**
     * The ActivityApi service constructor.
     * @property {module:api/ActivityApi}
     */
    ActivityApi: ActivityApi,
    /**
     * The ActivityIntradayTimeSeriesApi service constructor.
     * @property {module:api/ActivityIntradayTimeSeriesApi}
     */
    ActivityIntradayTimeSeriesApi: ActivityIntradayTimeSeriesApi,
    /**
     * The ActivityTimeSeriesApi service constructor.
     * @property {module:api/ActivityTimeSeriesApi}
     */
    ActivityTimeSeriesApi: ActivityTimeSeriesApi,
    /**
     * The AuthApi service constructor.
     * @property {module:api/AuthApi}
     */
    AuthApi: AuthApi,
    /**
     * The BodyAndWeightApi service constructor.
     * @property {module:api/BodyAndWeightApi}
     */
    BodyAndWeightApi: BodyAndWeightApi,
    /**
     * The BodyAndWeightTimeSeriesApi service constructor.
     * @property {module:api/BodyAndWeightTimeSeriesApi}
     */
    BodyAndWeightTimeSeriesApi: BodyAndWeightTimeSeriesApi,
    /**
     * The DevicesApi service constructor.
     * @property {module:api/DevicesApi}
     */
    DevicesApi: DevicesApi,
    /**
     * The FoodAndWaterApi service constructor.
     * @property {module:api/FoodAndWaterApi}
     */
    FoodAndWaterApi: FoodAndWaterApi,
    /**
     * The FoodAndWaterTimeSeriesApi service constructor.
     * @property {module:api/FoodAndWaterTimeSeriesApi}
     */
    FoodAndWaterTimeSeriesApi: FoodAndWaterTimeSeriesApi,
    /**
     * The FriendsApi service constructor.
     * @property {module:api/FriendsApi}
     */
    FriendsApi: FriendsApi,
    /**
     * The HeartRateIntradayTimeSeriesApi service constructor.
     * @property {module:api/HeartRateIntradayTimeSeriesApi}
     */
    HeartRateIntradayTimeSeriesApi: HeartRateIntradayTimeSeriesApi,
    /**
     * The HeartRateTimeSeriesApi service constructor.
     * @property {module:api/HeartRateTimeSeriesApi}
     */
    HeartRateTimeSeriesApi: HeartRateTimeSeriesApi,
    /**
     * The SleepApi service constructor.
     * @property {module:api/SleepApi}
     */
    SleepApi: SleepApi,
    /**
     * The SubscriptionsApi service constructor.
     * @property {module:api/SubscriptionsApi}
     */
    SubscriptionsApi: SubscriptionsApi,
    /**
     * The UserApi service constructor.
     * @property {module:api/UserApi}
     */
    UserApi: UserApi
  };

  return exports;
}));
